ThermodynamicState support for more barostats

Olllom opened the pull request:
## Description
Thermodynamic state supports
- MonteCarloAnistropicBarostat with same pressure along all axes
- MonteCarloMembraneBarostat with zero surface tension

see #436

## Todos
- [ ] Implement feature / fix bug
- [ ] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [ ] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [ ] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst)Notable points that this PR has either accomplished or will accomplish.

## Status
- [ ] Ready to go

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1235,9 +1241,12 @@ def _initialize(self, system, temperature=None, pressure=None):
+            self._pressure = self._get_barostat_pressure(barostat)
         else:
             self._pressure = pressure  # Pressure here can also be None.
+        self._barostat_type = openmm.MonteCarloBarostat
```
What do you think about making `_barostat_type` a private property that simply returns `type(self.barostat)`? Because this attribute is accessed very rarely, I'd prefer to make it slightly slower but avoid redundancies.

Olllom replied:
I would rather like to keep it as it is. The reason is that `_initialize` calls `_is_barostat_consistent` indirectly (through `_unsafe_set_system` and `_check_system_consistency`_). At this point, `self.barostat` is not available, since `self._standard_system` is not initialized, yet.

andrrizzi replied:
I see what you mean. It may be safe after all though because, if I understand correctly the code I wrote, `_is_barostat_consistent` shouldn't be called when `self.barostat` is unavailable (I'm looking at these lines: https://github.com/Olllom/openmmtools/blob/other_barostats/openmmtools/states.py#L1341-L1342).

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1573,6 +1582,19 @@ def _find_barostat(cls, system, get_index=False):
+                scaled = [barostat.getScaleX(), barostat.getScaleY(), barostat.getScaleY()]
+                if sum(scaled) == 0:
+                    raise ThermodynamicsError(ThermodynamicsError.UNSUPPORTED_ANISOTROPIC_BAROSTAT)
+                active_pressures = [pressure for pressure, active in zip(pressures, scaled)]
```
`active` here is unused

Olllom replied:
Good catch! Thank you.

andrrizzi commented:
As an extension (but feel free to postpone this to a future PR if you don't need it) we could allow `self.pressure` to be a vector to support different values of pressures for the axes of the anisotropic MC barostat.

Olllom commented:
An easier and more relevant addition for us would be nonzero surface tension. I will add some code for this to the pull request.

andrrizzi commented:
Good point about the pV work. Feel very free to implement only the bits you need. This is already very helpful.

Olllom commented:
I added
1. support for non-zero surface tension
2. support for setting the platform properties in the context cache and in the thermodynamic state

Point 2. should enable multistate simulations on CUDA in mixed and double precision,
by binding the MCMC moves to caches created with the 'precision'

It also makes it unnecessary to reduce CUDA_VISIBLE_DEVICES to a single device in multinode-gpu simulations. Instead the device ID can be set as a platform property.

Olllom commented:
This PR is ready to go on my end.

Olllom commented on openmmtools/states.py:
```diff
@@ -2032,6 +2105,11 @@ def volume(self):
         return _box_vectors_volume(self.box_vectors)
 
+    @property
+    def area(self):
```
Done

Olllom commented on openmmtools/cache.py:
```diff
@@ -447,13 +453,17 @@ def __getstate__(self):
             self._platform = None
         else:
             self._platform = openmm.Platform.getPlatformByName(serialization['platform'])
+        if not 'platform_properties' in serialization:
```
Done

Olllom commented on openmmtools/cache.py:
```diff
@@ -303,8 +306,11 @@ class ContextCache(object):
+    def __init__(self, platform=None, platform_properties=None, **kwargs):
         self._platform = platform
+        self._platform_properties = platform_properties
+        if platform_properties is not None and platform is None:
```
Done

Olllom commented on docs/releasehistory.rst:
```diff
@@ -1,6 +1,11 @@
+Enhancements
+------------
+- Added support for anisotropic and membrane barostats in `ThermodynamicState`
+- Added support for platform properties in ContextCache (e.g. for mixed and double precision CUDA in multistate sampler)
```
Done

Olllom commented:
Hi Andrea,

thanks for your suggestions and sorry for the delay. I have incorporated all of your requests. 

In order to allow compatibility between states with different surface tensions, I basically mimicked the pressure implementation. This should be the most consistent way to standardize the system, since the surface tension needs to be defined by the ThermodynamicState rather than the system barostat.

Instead of having a field for the barostat type, I added a private method `_is_barostat_type_consistent`. This should avoid the inconsistencies you mentioned above.

jchodera commented:
Since this PR has added significant complexity, is there anything we could implement at the OpenMM API  for these barostats that would significantly simplify this code? We're looking for features to add for the next OpenMM release, so if there is an easy way to extend the OpenMM API to facilitate this, please let us know!

andrrizzi commented:
Thanks @Olllom ! I'll take a look at the changes before the end of the week. Now that we merged the multiple alchemical region features, I should be able to take a look at your other PR modifying the alchemical factory too next week.

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1303,9 +1315,15 @@ def _initialize(self, system, temperature=None, pressure=None):
-        if barostat is not None:
-            self._barostat_type = type(barostat)
+
+        # If surface tension is None, we try to infer the pressure from the barostat.
```
"infer the pressure" -> "infer the surface tension"

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1757,6 +1804,32 @@ def _set_barostat_temperature(barostat, temperature):
+
+
+    def _set_barostat_surface_tension(self, barostat, gamma):
+        # working around a bug in the unit conversion https://github.com/openmm/openmm/issues/2406
```
Nice catch!

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1757,6 +1804,32 @@ def _set_barostat_temperature(barostat, temperature):
+        # working around a bug in the unit conversion https://github.com/openmm/openmm/issues/2406
+        if isinstance(gamma, unit.Quantity):
+            gamma = gamma.value_in_unit(unit.bar * unit.nanometer)
+        # barostat = self._find_barostat(system)
```
Probably forgotten here.

andrrizzi commented on openmmtools/states.py:
```diff
@@ -1618,6 +1648,11 @@ def set_temp(_integrator):
     _SUPPORTED_BAROSTATS = {'MonteCarloBarostat', 'MonteCarloAnisotropicBarostat', 'MonteCarloMembraneBarostat'}
 
+    @property
+    def _barostat_type(self):
```
Probably forgotten here since it's unused.

andrrizzi commented on openmmtools/tests/test_states.py:
```diff
@@ -437,7 +437,7 @@ def test_property_system(self):
                       (self.multiple_barostat_alanine, TE.MULTIPLE_BAROSTATS),
                       (self.unsupported_anisotropic_barostat_alanine, TE.UNSUPPORTED_ANISOTROPIC_BAROSTAT),
-                      (self.supported_anisotropic_barostat_alanine, TE.INCONSISTENT_BAROSTAT),
+                      #(self.supported_anisotropic_barostat_alanine, TE.INCONSISTENT_BAROSTAT),
```
Did you mean to comment this test case?

Olllom commented:
Great!
All fixed.

andrrizzi commented:
Almost! It looks there's one of the new tests failing: https://travis-ci.org/choderalab/openmmtools/jobs/603266322#L3314

andrrizzi commented:
There are also these other two errors in the "allowed failures" that I haven't seen before, but I'm not sure this is related to your changes so I'd address in a separate PR unless you know why that is happening: https://travis-ci.org/choderalab/openmmtools/jobs/603266325#L3185

andrrizzi commented:
I confirm that the latter tests are unrelated as they appear also in a PR without these changes so there's just the surface tension test to fix.

jchodera commented:
Is there anything we could implement at the OpenMM API for these barostats that would significantly simplify this code?

andrrizzi commented:
>Is there anything we could implement at the OpenMM API for these barostats that would significantly simplify this code?

Nothing particularly useful comes to mind. I think most of the complexity here comes from the fact that we're providing a unified API to a bunch of objects that have different parameters and (consequently) different APIs. We could create a general `Barostat` Python object in OpenMM that does these checks for us, but I think the total complexity (i.e., OpenMM + OpenMMTools) will be identical in the end.

Olllom commented:
@jchodera Something like this old PR of yours would be useful to simplify the API: https://github.com/openmm/openmm/issues/1440 

In this way, OpenMM could provide a simplified API to set things like temperature, pressure, surface tension, but also step size, etc. But this would only simplify the present code to a certain degree. IMHO a lot of the complexity in the `ThermodynamicState` is due to the `_standardized_system`, from which the data stored explicitly in the `ThermodynamicState` (pressure, temperature, surface tension) is excluded. However, I do not see a good way around this without hurting performance.

On a side note, one rather boring feature that would be very useful for us is native OpenMM support for the van-der-Waals force switch (see https://github.com/choderalab/openmmtools/pull/431). We have to use custom forces for the vdW in most of our simulations to get the correct lipid areas, which is quite a significant performance hit.

Olllom commented:
@andrrizzi should be fixed now.

andrrizzi commented:
Fantastic. Thanks again for the really nice work on tests too. Merging.
